tools/summonerd/src/coordinator.rs

```
@@ -1,90 +1,69 @@
+            tracing::debug!(
+                participant_count = participant_count,
+                "top of coordinator loop"
+            );
+            // 1. Inform all participants of their position
+            self.queue.inform_all().await?;
+            // 2. Select a contributor (we may have to sleep repeatedly till the queue gets at
+            //    least one member).
+            let (contributor, _) = loop {
+                if let Some(out) = self.queue.prune_and_pop().await {
+                    break out;
                 }
-                Some((w, p)) => (w, p),
+                tokio::time::sleep(Duration::from_secs(QUEUE_SLEEP_TIME_SECS)).await;
             };
-            tracing::debug!(?who, "waiting for contribution");
-            self.contribute::<P>(who, participant).await?;
-            self.done_contribution_tx.send(()).await?;
+            // 3. Get their contribution, or time out.
+            self.contribute::<P>(contributor).await?;
         }
     }
 
-    #[tracing::instrument(skip(self, participant))]
-    async fn contribute<P: Phase>(
-        &mut self,
-        contributor: Address,
-        participant: Participant,
-    ) -> Result<()> {
+    #[tracing::instrument(skip_all, fields(address = ?contributor.address()))]
+    async fn contribute<P: Phase>(&mut self, contributor: Participant) -> Result<()> {
+        let address = contributor.address();
         match tokio::time::timeout(
             Duration::from_secs(P::CONTRIBUTION_TIME_SECS),
-            self.contribute_inner::<P>(contributor, participant),
+            self.contribute_inner::<P>(contributor),
         )
         .await
         {
             Ok(Ok(_)) => Ok(()),
             Err(_) => {
                 tracing::info!("STRIKE (timeout)");
-                self.storage.strike(&contributor).await?;
+                self.storage.strike(&address).await?;
                 Ok(())
             }
             Ok(Err(e)) => Err(e),
```

Why do we want to bubble up an error here if `contribute_inner` fails? If a participant does something bad (e.g., submitting something that fails validation), won't this still crash the server?